Update rmt translator

docbacardi opened the pull request:
Fixes #3628 .

- [x] This PR is for the `dev-esp32` branch rather than for the `release` branch.
- [x] This PR is compliant with the [other contributing guidelines](/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/*`.

This set of patches adds optional parameters to the WS2812 module to set the duration of the reset signal and the duration for the high and low phase of the 0 and 1 symbols. This allows to control many other RGB LEDs besides the WS2812.
The main intention of this patch is to control the RGB LED of the ESP32-C3 dev kit.

docbacardi committed:
Choose the number of RMT buffers in the ws2812 module.

The number of buffers required for optimal operation should be selected
by the ws2812 module, not the caller.

docbacardi committed:
Add parameters for RGB LED bit times.

This patch adds compatibility for different RGB LEDS besides the WS2812.
ESP evaluation boards like the ESP32-C3-DevKitM-1 use an SK68XXMINI-HS
RGB LED which does not respond to the timings of this module.
The patch adds optional parameters for the bit timings to the write
function. If the new parameters are not supplied, the old values are used.
An example for the SK68XXMINI-HS is provided in the documentation.

docbacardi committed:
Remove restrictions from RTM translator.

The old RMT translator was not able to split the bits of the source
data into the size requested by the RMT transmitter. Either all 8 bits
of an input byte were translated or none.
The new routine removes the restriction by delivering exactly the
requested amount of data to the transmitter, which results in a more
balanced buffering of translated data under load.

docbacardi committed:
Add a parameter for the RGB LED reset time.

This patch introduces a new optional parameter for the reset time
in the RGB LED communication. The default is 51.2 microseconds. A
value of 0 sends no reset signal, which allows a small optimisation
for consecutive write commands.

Please note that the reset time of the old code should be 50
microseconds, as the define WS2812_DURATION_RESET suggested. Due to the
restrictions of the old RMT translator routine, it was slightly
increased to 51.2 microseconds. This patch keeps the value of 51.2
microseconds to be as compatible as possible.

docbacardi committed:
Minimize the time drift between RMT channels.

Place all RMT channels in a group to minimize the time drift between
the signals. Please note that this feature is not available on all
platforms.

pjsg commented on components/platform/ws2812.c:
```diff
@@ -32,112 +32,180 @@
+        /* Get the current byte. */
+        ucData = pucData[cnt_in] << ucBitPos;
+
+        while( cnt_in<src_size && cnt_out<wanted_num )
```
I suspect that there is an issue here if the `cnt_out<wanted_num` fails when not on a byte boundary. The original code validated that the entire byte would fit before it tried to encode it. There isn't a way to signal back to the rmt library that only part of a byte has been encoded.

docbacardi replied:
There is a good way to signal back that a byte was not completely processed: do not increase ```translated_size```.
The output parameter ```translated_size``` of the translator function reports how many bytes were processed. If a byte was not completely processed, it should not be considered in ```translated_size```. Additional information for the decoding process like the bit position can be stored in the translator context. Please note that the number of consumed raw bytes and the number of produced ```rmt_item32_t``` elements are not forced to have a fixed relationship of 1:8. The documentation has a note that it should be like this, but it is not a strict requirement (see [here](https://docs.espressif.com/projects/esp-idf/en/v4.3/esp32/api-reference/peripherals/rmt.html#_CPPv415sample_to_rmt_t)). The current routine inserts a reset signal with 8 ```rmt_item32_t``` elements without consuming a single raw byte for it (see [here](https://github.com/nodemcu/nodemcu-firmware/blob/10e378394861b8261eac4366658dad069aa12f63/components/platform/ws2812.c#L100)), and it is working apparently well.

I'm afraid my commit messages did not give a good explanation why I would like to change an implementation which is doing a good job. Playing around with the translator callback I found out that the transmission of RMT pulses stops if the ```wanted_num``` parameter is not met.

Example: I want to send 12 raw bytes which would give 96 ```rmt_item32_t``` elements after translation. The first call to the translator function has ```wanted_num``` set to 48, but my test code decides to provide only 40. The transmission stops after these pulses. The translator function is never called for the remaining 56 bytes of raw data. This was tested with an ESP32-C3.

Fortunately all observed values for the ```wanted_num``` parameter were multiples of 8. But I did not find a guarantee for it. That's why I changed the translator function. Am I too paranoid?

pjsg replied:
Thanks -- I missed the saving of the bitpos in the context.

pjsg commented on docs/modules/ws2812.md:
````diff
@@ -44,6 +52,10 @@ ws2812.write({pin =  4, data = string.char(255, 0, 0, 255, 0, 0)},
 ```
 
+```lua
+ws2812.write({pin = 8, reset = 800, t0h = 3, t0l = 9, t1h = 6, t1l = 6, data = string.char(1, 0, 0)}) -- turn the SK6812 RGB led on the ESP32-C3-DevKitM-1 to green
````
Does this work -- I would have thought that using 255 as the intensity rather than 1. Also, what is the color ordering for the SK6812? You set the first value to get green, which implies that it isn't RGB.

docbacardi replied:
The SK6812 is a very bright LED. A value of 1 is clearly visible. Setting it to 255 was painful for my eyes.

You are correct, the color ordering is really green - red - blue, see end of page 6 in the data sheet [here](https://cdn-shop.adafruit.com/product-files/1138/SK6812+LED+datasheet+.pdf). Thanks, the patch is updated.

However the WS2812 has the same color ordering (see [here](https://cdn-shop.adafruit.com/datasheets/WS2812.pdf), page 5). The APA104 is the only LED in the list using RGB color ordering (see [here](https://www.ledlightinghut.com/files/APA104%20Datasheet.pdf), page5). The current [documentation](https://github.com/nodemcu/nodemcu-firmware/blob/10e378394861b8261eac4366658dad069aa12f63/docs/modules/ws2812.md?plain=1#L14) uses the term "RGB" or "RGBW" for all of them:
```
Send data to up to 8 led strip using its native format which is generally Green,Red,Blue for RGB strips and Green,Red,Blue,White for RGBW strips.
```

If you agree that this is a bit confusing I will open a new ticket for it.

pjsg replied:
Yeah -- that makes sense

docbacardi committed:
Fix the description of the SK6812 LED in the example code.

The SK6812 expects the data for the green LED first, then red and
finally blue. It should be described as a GRB LED.
